README.md

```diff
@@ -2,18 +2,126 @@
 
-[![Clojars Project](https://img.shields.io/clojars/v/com.github.clojure-lsp/lsp4clj-protocols.svg)](https://clojars.org/com.github.clojure-lsp/lsp4clj-protocols)
+To initialize a server that will read from stdin and write to stdout:
 
-The lsp4clj-protocols contains only the protocols/interfaces for servers that want to extend the official LSP protocol to provide more features, also, it is used by lsp4clj-server itself.
+```clojure
+(lsp4clj.server/stdio-server {:in System/in, :out System/out})
+```
 
-## Known LSPs users
+The returned server will have a core.async `:log-ch`, from which you can read server logs (vectors beginning with a log level).
+
+```clojure
+(async/go-loop []
+  (when-let [[level & args] (async/<! (:log-ch server))]
+    (apply logger/warn level args)
+    (recur)))
+```
+
+### Receive messages
+
+To receive messages from a client, lsp4clj defines a pair of multimethods, `lsp4clj.server/receive-notification` and `lsp4clj.server/receive-request` that dispatch on the method name (as defined by the LSP spec) of an incoming JSON-RPC message.
+
+Server implementors should create `defmethod`s for the messages they want to process. (Other methods will be logged and responded to with a generic "Method not found" response.)
+
+These `defmethod`s receive 3 arguments, the method name, a "context", and the `params` of the [JSON-RPC request or notification object](https://www.jsonrpc.org/specification#request_object). The keys of the params will have been converted (recursively) to kebab-case keywords. Read on for an explanation of what a "context" is and how to set it.
+
+```clojure
+;; a notification; return value is ignored
+(defmethod lsp4clj.server/receive-notification "textDocument/didOpen"
+  [_ context {:keys [text-document]}]
+  (handler/did-open context (:uri text-document) (:text text-document))
+  
+;; a request; return value is converted to a response
+(defmethod lsp4clj.server/receive-request "textDocument/definition"
+  [_ context params]
+  (->> params
+       (handler/definition context)
+       (conform-or-log ::coercer/location)))
+```
+
+The return value of requests will be converted to camelCase json and returned to the client. If the return value looks like `{:error ...}`, it is assumed to indicate an error response, and the `...` part will be set as the `error` of a [JSON-RPC error object](https://www.jsonrpc.org/specification#error_object). It is up to you to conform the `...` object (by giving it a `code`, `message`, and `data`.) Otherwise, the entire return value will be set as the `result` of a [JSON-RPC response object](https://www.jsonrpc.org/specification#response_object). (Message ids are handled internally by lsp4clj.)
+
+### Send messages
+
+Servers also initiate their own requests and notifications to a client. To send a notification, call `lsp4clj.server/send-notification`.
+
+```clojure
+(->> {:message message
+      :type type
+      :extra extra}
+     (conform-or-log ::coercer/show-message)
```

Comment:
Why require conform? can't lsp4clj do the conform on its side as it knows the method and to what conform?

Reply:
That's true, lsp4clj has enough information to do conformation automatically. I guess the main reason I didn't do it that way is that I don't like the conformation, and so I wanted to make it optional.

In the particular case of `::coercer/show-message`, the reason I don't like the conformation is that it is a very complicated way to convert `:type` from a keyword like `:warning` to an enum value like `2`. Otherwise, it serves no purpose, IMO. Beyond `::coercer/show-message`, I feel the same way about most of the conformation. To me, the conformation is a relic of lsp4j. It was a sane way to create Java objects, but it is too convoluted for doing simple translations like `:warning` -> `2`.

I started to abandon the conformation but then decided that it'd add too much noise to an already big change set. If you'd like, we can work on moving the conformation back to the server, but I'm inclined to do that in a separate PR.

This is another place where having a "route" object could be handy. You could add a `:route/conform-output?` flag which indicated to the server that it should do its default conformation. That would let clojure-lsp opt-out.

Reply:
I see, most of conformation is not needed anymore indeed, we can think in the future replacing with safe checks/adapters and only that probably, it'd make project easier to understand too I guess

Reply:
Yep, that's my hope. We'll see though... There's rarely any reason to change the existing specs, so they might stick around.